### PR TITLE
Add Scalarization Patterns for `AtenToDtypeOp`, `AtenNegOp`, `AtenRemainderTensorOp`

### PR DESCRIPTION
**Pull request by zjgarvey**
1. adds a lowering for `aten.neg.int` and `aten.remainder.int` to arith.
2. adds a scalarization pattern for `aten.neg` and `aten.remainder.Tensor` ops.
3. improves folding of `aten.mul.int`
4. adds a scalarization pattern for `aten.to.dtype` which relies on scalar cast ops and basic C++ casting between `double` and `int64_t`.
5. improves rank-0 case handling for `FoldAtenSplatPattern`
6. removes a bug with `aten.unflatten.int` decomposition incorrectly generating a constant size int from a dynamic shape.
7. simplifies the dim list for `aten.unflatten.int` ops generated from the `aten.view` canonicalization in scalarize shapes. 

All of these changes were necessary to unblock <https://github.com/iree-org/iree/issues/18899>.

### CODE REVIEW

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 1]
```diff
@@ -4068,6 +4068,10 @@ OpFoldResult AtenMulIntOp::fold(FoldAdaptor adaptor) {
   int64_t lhs, rhs;
   bool lConstant = matchPattern(getOperand(0), m_TorchConstantInt(&lhs));
   bool rConstant = matchPattern(getOperand(1), m_TorchConstantInt(&rhs));
+  if ((lConstant && lhs == 1))
```
Comment (rsuderman) [inline lib/Dialect/Torch/IR/TorchOps.cpp:4071]: nit: Remove extra set of brackets

(and below too)

File: lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp [hunk 2]
```diff
@@ -746,6 +746,98 @@ class PropagateAtenItemPattern : public OpRewritePattern<AtenItemOp> {
 };
 } // namespace
 
+namespace {
+
+LogicalResult convertOpFoldResults(ImplicitLocOpBuilder &b,
+                                   SmallVector<OpFoldResult> &converted,
+                                   SmallVector<OpFoldResult> &elements,
+                                   Type inputDtype, Type resultDtype) {
+  auto inputIsInt = dyn_cast<mlir::IntegerType>(inputDtype);
```
Comment (rsuderman) [inline lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp:755]: Probably should cache similarly for `fp` so the checks below just check each global.

File: lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp [hunk 3]
```diff
@@ -746,6 +746,98 @@ class PropagateAtenItemPattern : public OpRewritePattern<AtenItemOp> {
...
+  auto resultIsInt = dyn_cast<mlir::IntegerType>(resultDtype);
+  if (!inputIsInt && !isa<mlir::FloatType>(inputDtype))
+    return failure();
+  if (!resultIsInt && !isa<mlir::FloatType>(resultDtype))
+    return failure();
+  // if dtypes are both int or both float, no conversion needed
+  if (static_cast<bool>(inputIsInt) == static_cast<bool>(resultIsInt)) {
```
Comment (rsuderman) [inline lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp:762]: Oh man, I have a wonder trick. Xor.

`if (!(inputIsInt ^ resultIsFloat))`

File: lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp [hunk 4]
```diff
@@ -746,6 +746,98 @@ class PropagateAtenItemPattern : public OpRewritePattern<AtenItemOp> {
...
+    converted = elements;
+    return success();
+  }
+  for (auto e : elements) {
+    auto eValue = dyn_cast<Value>(e);
+    if (eValue && resultIsInt) {
```
Comment (rsuderman) [inline lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp:768]: <optional> Can we invert the conditionals and the loop? Thought is its a bit cleaner to see loops either converting everything to ints or everything to floats. Just a personal preference knowing that I believe it should be possible.

File: lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp [hunk 5]
```diff
@@ -841,6 +933,48 @@ class PropagateAtenArithmeticPattern : public OpRewritePattern<OpTy> {
...
+    SmallVector<OpFoldResult> resultFolds;
+    for (uint64_t i = 0; i < selfVals.size(); i++) {
+      resultFolds.push_back(
+          b.createOrFold<ScalarOpTy>(selfVals[i].getType(), selfVals[i]));
+    }
+    SmallVector<Value> resultVals;
+    if (failed(materializeFolds(b, resultFolds, resultVals)))
+      return failure();
+
+    if (resultTy.getSizes().size() == 0) {
```
Comment (rsuderman) [inline lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp:966]: I think we can do

`if(resultTy.getSizes().empty())`

File: lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp [hunk 6]
```diff
@@ -915,19 +1049,22 @@ class FoldAtenTensorSplatPattern : public OpRewritePattern<AtenTensorOp> {
     auto resultTy = cast<BaseTensorType>(op.getType());
     if (!resultTy.hasSizes() || !resultTy.areAllSizesKnown())
       return rewriter.notifyMatchFailure(op, "dynamic output shape");
+    if (resultTy.getSizes().size() == 0) {
```
Comment (rsuderman) [inline lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp:1052]: Empty comment above

**Review by rsuderman — CHANGES_REQUESTED** [review]

File: lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp (hunk 3 above)
Comment (zjgarvey) [reply lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp:762]: Ah, nice. Better than the ugly static casts.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 7]
```diff
@@ -4587,6 +4587,11 @@ class DecomposeAtenUnflattenIntOp
     if (!isValidDim(dimInt, inputRank))
       return rewriter.notifyMatchFailure(op, "dim is not a valid dim");
 
+    if (inputShape[dimInt] == Torch::kUnknownSize &&
+        llvm::count(sizesInts, -1) > 0)
+      return rewriter.notifyMatchFailure(
+          op, "Unimplmented: dynamic unflatten dim with an inferred size.");
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:4593]: Typo here: `Unimplmented` -> `Unimplemented`

File: lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp [hunk 8]
```diff
@@ -1182,8 +1337,29 @@ class CanonicalizeAtenViewPattern : public OpRewritePattern<AtenViewOp> {
...
+      int64_t dynCount = 0;
+      int64_t dynIdx = 0;
+      for (auto [i, v] : llvm::enumerate(unflattenSizes)) {
+        int64_t szeInt;
+        if (!matchPattern(v, m_TorchConstantInt(&szeInt))) {
+          dynCount++;
+          dynIdx = i;
+          continue;
+        }
+        // if we a -1 already, make dynCount invalid and break
```
Comment (vivekkhandelwal1) [inline lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp:1352]: Another typo

**Review by vivekkhandelwal1 — CHANGES_REQUESTED** [review]

File: lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp (hunk 2 above)
Comment (zjgarvey) [reply lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp:755]: I think I'm going to skip this and keep the inlined `isa` checks.

File: lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp (hunk 3 above)
Comment (zjgarvey) [reply lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp:762]: I still need the `static_cast<bool>`'s apparently, so I'm going to leave this as-is.

**Review by rsuderman — APPROVED** [review]

**Pull request merged**